cassiopeia/solver/ilp_solver_utilities.pyx

```diff
@@ -1,22 +1,157 @@
+        f"(Process: {pid}) Estimating a potential graph with "
+        "a maximum layer size of "
+        f"{maximum_potential_graph_layer_size} and a maximum "
+        f"LCA height of {lca_height}."
+    )
+
+    cdef int effective_threshold
+    cdef int max_layer_width
+
+    cdef list layer_sizes = []
+    cdef list distance_thresholds = []
+    cdef list previous_layer_edges = []
+    cdef list current_layer_edges = []
+
+    cdef int n_characters = character_array.shape[1]
+
+    cdef int distance_threshold = 0
+
+    # Cast character states to strings.
+    # Also, because we've converted states into strings, missing data items
+    # that were negative numbers will take up two characters. Thus,
+    # we convert missing data to a universal character "-"
+    character_states = np.array(
+        [
+            "|".join(arr).replace(str(missing_state_indicator), "-")
+            for arr in np.array(character_array)
+        ]
+    )
+
+    while distance_threshold < (lca_height + 1):
+
+        current_layer_edges = []
+
+        source_nodes = character_states
+        effective_threshold = distance_threshold
+        max_layer_width = 0
+
+        while source_nodes.shape[0] > 1:
+
+            if source_nodes.shape[0] > maximum_potential_graph_layer_size:
+                logging.info(
+                    f"(Process: {pid}) Maximum layer size "
+                    "exceeded, returning network."
+                )
+
+                return previous_layer_edges
+
+            (next_layer, layer_edges) = infer_layer_of_potential_graph(
+                source_nodes, effective_threshold
+            )
+
+            # subset to unique values
+            if next_layer.shape[0] > 0:
+                next_layer = np.unique(next_layer)
+                # unique_idx = fast_unique(np.array([node.split("|") for node in next_layer]).astype(int))
+                # next_layer = next_layer[unique_idx]
+
+            if (
+                next_layer.shape[0] > maximum_potential_graph_layer_size
+                and len(previous_layer_edges) > 0
+            ):
+                return previous_layer_edges
+
+            # edges come out as rows in a numpy matrix, where the first
+            # (2*n_characters-1) positions correspond to the parent and the
+            # last (2*n_characters) positions correspond to the child
+            layer_edges = [
+                (
+                    "|".join(e.split("|")[:n_characters]),
+                    "|".join(e.split("|")[n_characters:]),
+                )
+                for e in layer_edges
+                if "|".join(e.split("|")[:n_characters])
+                != "|".join(e.split("|")[n_characters:])
+            ]
+            current_layer_edges += layer_edges
+
+            if source_nodes.shape[0] > next_layer.shape[0]:
+                if effective_threshold == distance_threshold:
+                    effective_threshold *= 3
+            source_nodes = next_layer
```

Comment:
remind me, is there a reason behind the 3? Or is it just arbitrary

Reply:
No reason -- this was just in the original codebase that Alex and I developed a few years ago. I think Alex chose this modifier because it worked well empirically